### DISCUSSION TITLE
Composer still using Mac OS PHP version after php@7.4 Brew install.

### DISCUSSION THREAD

**Discussion by rafaqf** [shivammathur/homebrew-php#359] [Q&A / Help]
Hello everyone,

I've followed the steps to update my Mac OS PHP version to 7.4 as my composer is requiring it and not letting me commit. Unfortunately, I can't commit yet. 

This is how my PHP looks:

```
RQF@MacBook-Pro ~ % find / -name php -type f -perm +111 2>nul
/usr/bin/php
/usr/local/Cellar/php@7.4/7.4.20/bin/php
/System/Volumes/Data/usr/local/Cellar/php@7.4/7.4.20/bin/php

RQF@MacBook-Pro ~ % whereis php
/usr/bin/php

RQF@MacBook-Pro ~ % which php
/usr/local/opt/php@7.4/bin/php

RQF@MacBook-Pro ~ % php -v
PHP 7.4.20 (cli) (built: Jun  4 2021 01:49:56) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.20, Copyright (c), by Zend Technologies
```

Everything looks ok, as you can see, but my Mac OS PHP still points to the old one instead of to the new one installed by Brew. Or at least, Composer is looking to the old one - which should be fixed after installing and linking the new php@7.4 - and I have no idea what else should I try to fix that.

Thanks a lot in advance, any suggestion is more than welcome!

**Comment by shivammathur** [comment]
`composer` uses `env` to find php in PATH.

- Run `echo $PATH`. 
- `/usr/local/bin` should be before `/usr/bin` in `PATH`.
- If It is not, edit `/etc/paths` file and make sure `/usr/local/bin` is above `/usr/bin`. Make sure you restart the terminal after editing this file.

**Reply by rafaqf** [reply]
@shivammathur This is the output:

```
RQF@MacBook-Pro ~ % echo $PATH
/usr/local/sbin:/usr/local/opt/php@7.4/sbin:/usr/local/opt/php@7.4/bin:/Users/RQF/.nvm/versions/node/v12.16.0/bin:/usr/local/opt/php@7.4/bin/php:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin
```

Thanks for your answer!

**Reply by rafaqf** [reply]
this is my `/etc/paths`:

```
/usr/local/opt/php@7.4/bin/php
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

**Reply by shivammathur** [reply]
In that case, `composer` should use the brewed PHP.

**Reply by rafaqf** [reply]
unfortunately, it is not, and I have no idea why. Do you have any suggestions?

**Reply by shivammathur** [reply]
what is the output of `env php -v`

**Reply by shivammathur** [reply]
Also, check if composer in PATH is actually a symlink to the system PHP calling composer, in that case, remove the symlink.

**Reply by rafaqf** [reply]
```
RQF@MacBook-Pro ~ % env php -v
PHP 7.4.20 (cli) (built: Jun  4 2021 01:49:56) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.20, Copyright (c), by Zend Technologies
```

**Reply by rafaqf** [reply]
> Also, check if composer in PATH is actually a symlink to the system PHP calling composer, in that case, remove the symlink.

sorry I don't know how to check that, if you could throw me the commands would be very helpful :)

**Reply by shivammathur** [reply]
The output of `env php -v` means composer should use PHP from brew.
Can you tell me the output of `type php` and `type composer`

**Reply by rafaqf** [reply]
```
RQF@MacBook-Pro ~ % type php
php is /usr/local/opt/php@7.4/bin/php
```

I use composer from my repository as `php composer.phar "command"`, so the composer type output is:

```
RQF@MacBook-Pro ~ % type composer
composer not found
```

**Reply by shivammathur** [reply]
In your repository. do you have a PHP symlink which links to the system PHP? You can check this using the command `ls -la` in your repo.

I would recommend installing composer globally using brew `brew install composer`.

**Reply by rafaqf** [reply]
No symlink as I can see, all folders and files. Just installed composer globally but from my repository, where I have to commit, the composer installed there is the one still being used. 🤦🏼

**Reply by shivammathur** [reply]
After installing composer globally do not run the local phar, run composer directly like `compose install` without the PHP prefix.

**Reply by rafaqf** [reply]
The problem is on git commands, that are using the composer from our repo. For example when trying to commit I get this error message.

`Fatal error: Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 7.4.0". You are running 7.3.24-(to be removed in future macOS). in /Library/WebServer/Documents/chesskid/vendor/composer/platform_check.php on line 24`

where `platform_check.php` created by composer in my repo looks like this:

```
<?php

// platform_check.php @generated by Composer

$issues = array();

if (!(PHP_VERSION_ID >= 70400)) {
    $issues[] = 'Your Composer dependencies require a PHP version ">= 7.4.0". You are running ' . PHP_VERSION . '.';
}

if ($issues) {
    if (!headers_sent()) {
        header('HTTP/1.1 500 Internal Server Error');
    }
    if (!ini_get('display_errors')) {
        if (PHP_SAPI === 'cli' || PHP_SAPI === 'phpdbg') {
            fwrite(STDERR, 'Composer detected issues in your platform:' . PHP_EOL.PHP_EOL . implode(PHP_EOL, $issues) . PHP_EOL.PHP_EOL);
        } elseif (!headers_sent()) {
            echo 'Composer detected issues in your platform:' . PHP_EOL.PHP_EOL . str_replace('You are running '.PHP_VERSION.'.', '', implode(PHP_EOL, $issues)) . PHP_EOL.PHP_EOL;
        }
    }
    trigger_error(
        'Composer detected issues in your platform: ' . implode(' ', $issues),
        E_USER_ERROR
    );
}
```

No idea how to solve that. This was supposed to fix the issue: `php composer.phar dump-autoload -o`, but the `platform_check` file doesn't update.

**Reply by shivammathur** [reply]
What is the output of `php -r "echo PHP_VERSION_ID;"`?

There is no issue with the `platform_check` file. The PHP binary used to execute this file is the system one, this seems to be hardcoded somewhere.

**Reply by rafaqf** [reply]
```
RQF@MacBook-Pro ~ % php -r "echo PHP_VERSION_ID;"
70420%
```

Thanks a lot for your help :)

**Reply by rafaqf** [reply]
I've found that in the `/usr/bin/php-config`, the version is `7.3.24`.

```
RQF@MacBook-Pro ~ % grep -Rw /usr -e "7.3.24"
Binary file /usr/bin/php matches
... (many with permission denied)
/usr/bin/php-config:version="7.3.24-(to be removed in future macOS)"
Binary file /usr/libexec/apache2/libphp7.so matches
Binary file /usr/sbin/php-fpm matches
grep: /usr/local/var/homebrew/linked/php@7.4/php5: Too many levels of symbolic links
```

Shouldn't that file be updated by homebrew?

**Reply by shivammathur** [reply]
No, brew symlinks binaries to `/usr/local/bin`.

**Reply by rafaqf** [reply]
So it seems that my repo composer is probably still checking that `php-config` or others, do you have any suggestion to fix that? Maybe I should delete all the macOS PHP files? I need to change some permissions in order to do that.

**Reply by shivammathur** [reply]
That might need disabling SIP and I would not recommend doing that. Your issue is related to how composer is run.
You can run composer with brew PHP `/usr/local/opt/php@7.4/bin/php composer ...`. This should work.
